5단계 - 자동차 경주(추가 연습)

ghtjr410 opened the pull request:
# 자동차 경주(추가 연습) 과제 제출
안녕하세요!
자동차 경주(추가 연습) 과제 제출합니다.
## 미션 진행 방향
현재 자신의 상태에 따라 미션을 새로 구현하거나 한번 더 리팩터링 해볼 것을 추천하셨는데, 저는 완전히 새롭게 구현하는 방향을 선택했습니다.
선택 이유
- 이미 구조화된 코드를 리팩터링하는 것은 4단계에서 충분히 연습
- 처음부터 다시 구현하면서 이전에 놓쳤던 설계 포인트를 발견할 수 있을 것이라 판단
- TDD 사이클을 더 자연스럽게 체득하고 싶었음
## 학습 목표
이번 미션에서 집중한 목표는 다음과 같습니다:
- TDD로 처음부터 끝까지 구현
- 객체지향 생활체조 원칙 준수
- 테스트 경계를 명확화하여 무엇을 테스트할지/하지 않을지 명확한 기준 수립
- 주생성자/부생성자 적극 활용 
- 이전 피드백 재발 방지
## 학습 결과
- 미션을 진행하며 "무엇을 테스트해야 하고, 무엇은 테스트하지 않아도 되는가"에 대한 저만의 기준을 세울 수 있었습니다.
- "하위 객체의 책임이 명확하고 테스트되었다면, 단순 위임은 테스트하지 않는다. 하지만 조건/분기/조합이 있다면 반드시 테스트한다."
- 이 기준 덕분에 과도한 테스트와 부족한 테스트 사이의 균형을 찾을 수 있었습니다.

리뷰 잘 부탁드립니다!

ghtjr410 committed:
refactor: 이동 정책 인터페이스 및 구현체 네이밍 정리

ghtjr410 committed:
docs: 레이싱카 미션 문서 업데이트 및 신규 리팩토링 문서 추가

docs/04-racing-car-winner.md
- PR 링크에 실제 GitHub PR URL 추가

docs/05-racing-car-refactoring.md
- 5단계 리팩토링용 신규 문서 생성
- TDD 사이클, 객체지향 원칙, 설계 목표 등 학습 목표 추가
- 미션 요구사항 및 체크리스트 작성
- 구현 기능 목록 섹션 생성

ghtjr410 committed:
refactor: 4단계까지의 기존 레이싱카 구현 전체 삭제

ghtjr410 committed:
docs: 레이싱 카 게임 3~4단계 기능 요구사항 추가

ghtjr410 committed:
feat: 자동차 이름 도메인 생성 및 유효성 검증 테스트 추가

CarName.java
- 자동차 이름을 표현하는 `CarName` 레코드 생성
- 이름 길이가 5자를 초과할 경우 예외를 발생하도록 검증 로직 추가
- 예외 메시지에 명확한 제약 조건 포함

CarNameTest.java
- 이름이 5자를 초과할 때 `IllegalArgumentException` 발생 여부 테스트 추가
- 예외 메시지 검증을 통해 정확한 오류 피드백 보장

ghtjr410 committed:
refactor: CarName 검증 로직 고도화 및 테스트 강화

CarName.java
- 최대 길이를 상수 `MAX_LENGTH`로 분리하여 관리 용이성 향상
- `validateNotBlank` 추가로 null/빈문자/공백 문자열 검증 강화
- `validateLength`로 길이 초과 검증 로직 분리
- 예외 메시지를 `String.format` 사용해 일관성 있게 변경

CarNameTest.java
- 정상 입력에 대한 파라미터 테스트 추가
- null, 빈 문자열, 공백 문자열에 대한 예외 검증 테스트 추가
- 길이 초과 검증 테스트 예외 메시지 정확히 검증하도록 수정

ghtjr410 committed:
feat: 자동차 이동 거리 도메인 생성 및 테스트 추가

CarDistance.java
- 자동차 이동 거리를 표현하는 `CarDistance` 레코드 신규 생성
- 기본 거리 값을 0으로 초기화하는 기본 생성자 추가
- `increase()` 메서드로 이동 시 거리 1 증가 기능 구현

CarDistanceTest.java
- 기본 생성 시 초기 값이 0인지 검증
- `increase()` 호출 시 거리 값이 1 증가하는지 테스트 추가

ghtjr410 committed:
feat: Car 도메인 추가 및 이름 기반 생성 기능 테스트 구현

Car.java
- 자동차를 구성하는 `CarName`, `CarDistance`를 포함한 `Car` 레코드 생성
- 문자열 이름만 받아도 자동으로 `CarName`과 초기 거리(`CarDistance`)를 생성하는 생성자 추가

CarTest.java
- 문자열 이름으로 `Car` 생성 시 예외가 발생하지 않는지 테스트 추가

ghtjr410 committed:
feat: Car 이동 기능 추가

Car.java
- 클래스 구조로 변경
- `move()` 메서드 추가해 조건 만족 시 거리 증가
- 이동 거리 조회 메서드(`getDistance`) 추가

MovableCondition.java
- 이동 가능 여부 판단 인터페이스 추가

CarTest.java
- Car 생성 테스트 추가
- 이동 조건 충족 시 거리 증가 테스트 추가

ghtjr410 committed:
refactor: Car 상태 조회 방식 개선

Car.java
- getDistance() 제거, status() 메서드로 CarStatus 반환

CarStatus.java
- Car 상태를 담는 신규 레코드 추가

CarTest.java
- status() 기반으로 거리 검증하도록 수정
- 이동 불가 테스트 추가

ghtjr410 committed:
feat: Cars 컬렉션 도메인 추가

Cars.java
- 문자열 입력으로 Car 목록을 생성하는 기능 추가
- 내부적으로 이름 분리 및 Car 객체 생성 로직 구현

CarsTest.java
- 문자열 입력 생성 시 예외 발생 여부 테스트 추가

ghtjr410 committed:
feat: Cars 생성자에 빈 리스트 검증 추가

Cars.java
- 입력 리스트가 null 또는 비어 있을 경우 예외 발생하도록 검증 로직 추가

CarsTest.java
- 빈 리스트/널 입력 시 예외 발생 테스트 추가

ghtjr410 committed:
feat: Cars 전체 이동 기능 추가

Cars.java
- 모든 Car 객체에 대해 move를 호출하는 moveAll() 메서드 추가

ghtjr410 committed:
feat: 레이싱 게임 진행 및 결과 조회 기능 추가

Cars.java
- 각 Car의 상태를 반환하는 statuses() 메서드 추가

RaceResult.java
- CarStatus 목록을 담는 결과 객체 추가

RacingGame.java
- 경주 진행(race) 및 결과 생성(result) 기능 구현

RacingGameTest.java
- 테스트 클래스 기본 구조 생성

ghtjr410 committed:
feat: CarDistance 비교 기능 추가

CarDistance.java
- max, isSame 비교 메서드 추가

CarDistanceTest.java
- 생성자 및 increase 테스트 수정
- max, isSame 파라미터화 테스트 추가

ghtjr410 committed:
feat: Car 생성자 확장 및 거리 비교 기능 추가

Car.java
- name, distance 직접 입력 가능한 생성자 추가
- 거리 비교용 maxDistance, isSameDistance 메서드 추가
- name 값 반환 메서드 추가

ghtjr410 committed:
feat: 우승자 판별 기능 추가

Cars.java
- 최대 거리 계산 및 Winners 반환 기능 추가

Winners.java
- 우승자 컬렉션 검증 포함한 레코드 생성

CarsTest.java
- winners 기능 테스트 추가

WinnersTest.java
- 생성자 검증 테스트 추가

ghtjr410 committed:
feat: 랜덤 이동 조건 기능 추가

RandomMovableCondition.java
- 랜덤값 기반 이동 가능 여부 판단 로직 추가
- RandomValueGenerator 주입 가능하도록 구성

RandomValueGenerator.java
- 랜덤값 생성 위한 함수형 인터페이스 추가

RandomMovableConditionTest.java
- 이동 가능/불가능 조건 파라미터화 테스트 추가

ghtjr410 committed:
feat: 레이싱 라운드 진행 로직 추가

RacingRound.java
- 총 라운드/현재 라운드 관리 클래스 추가
- 유효성 검증, 종료 여부 판단, next 진행 기능 구현
- equals/hashCode 재정의

RacingRoundTest.java
- 생성자 기본값 검증 테스트 추가
- 유효성 검증, isFinished, next 기능 테스트 추가

ghtjr410 committed:
feat: RacingGame에 라운드 기반 진행 및 종료 처리 추가

RacingGame.java
- RacingRound 기반으로 게임 진행 구조 변경
- race 메서드에 종료 여부 체크 및 예외 처리 추가
- winners, isFinished 기능 추가
- 생성자에 tryNumber 기반 라운드 초기화 추가

RacingGameTest.java
- 종료된 게임에서 race 호출 시 예외 발생 테스트 추가

ghtjr410 committed:
feat: CarStatus 출력 포맷 기능 추가

CarStatus.java
- 화면 출력용 formatForDisplay 메서드 추가

CarStatusTest.java
- formatForDisplay 결과 검증 테스트 추가

ghtjr410 committed:
feat: Winners 출력 포맷 기능 추가

Winners.java
- 우승자 이름 조합 후 출력하는 formatForDisplay 메서드 추가
- 우승자 이름 리스트 생성 메서드 추가

WinnersTest.java
- formatForDisplay 결과 검증 테스트 추가

ghtjr410 committed:
feat: 콘솔 입출력 기반 레이싱게임 실행 흐름 구현

Application.java
- 사용자 입력을 받아 RacingGame 실행
- 라운드 진행 루프 및 결과 출력 처리 추가

InputView.java
- 자동차 이름 및 시도 횟수 입력 기능 추가

ResultView.java
- 게임 진행 결과 및 우승자 출력 기능 추가
- CarStatus/Winners 출력 메서드 구현

ghtjr410 committed:
refactor: indent 2 우승자 조회 로직 스트림 기반으로 단순화

Cars.java
- winners 생성 방식 간소화
- findWinners 로직을 스트림 필터링 기반으로 개선
- 불필요한 리스트 생성 및 반복문 제거

ghtjr410 committed:
refactor: 이동 조건 관련 패키지 utils로 분리

ghtjr410 committed:
refactor: InputView 하드코딩된 메시지를 상수로 분리

InputView.java
- 자동차 이름/시도 횟수 안내 메시지를 상수로 추출
- 출력 문구 반복 제거로 가독성 개선

ghtjr410 committed:
docs: 레이싱카 리팩터링 요구사항에 구현 기능 목록 추가

javajigi commented on docs/05-racing-car-refactoring.md:
```diff
@@ -0,0 +1,91 @@
+## 코드 리뷰
+> PR 링크:
+> 
+## 나의 학습 목표
```
👍 
이런 목표를 설정하고 도전할 때 의식적인 연습이 되면서 빠르게 성장해 나감
일단 목표 설정만으로도 성공 가능성 높아보임

javajigi commented on src/main/java/racingcar/domain/CarDistance.java:
```diff
@@ -2,17 +2,21 @@
 
-    public CarDistance increase() {
-        return new CarDistance(this.value + MOVE_DISTANCE);
+    public int max(int that) {
+        return Math.max(this.value, that);
+    }
```
```suggestion
    public CarDistance max(int that) {

    }
    
    public CarDistance max(CarDistance that) {

    }
```
- 메서드 반환 값을 int가 아닌 CarDistance로 구현해 고려해 볼 요소 중 하나.
- max 메서드에 int 외에 CarDistance와 같은 타입을 받도록 오버로딩하는 것도 고려해볼 요소 중 하나.

이와 같이 여러 메서드를 오버로딩하면 이 메서드를 호출하는 개발자에게 편리성을 제공해 줄 수 있지 않을까?
한번씩 고려해 보면 좋겠음.

javajigi commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -1,64 +1,72 @@
-            throw new IllegalArgumentException("자동차 이름은 중복될 수 없습니다.");
-        }
+    public Cars(String inputNames) {
+        this(createCars(inputNames));
```
```suggestion
    public Cars(String inputNames) {
        this(createCars(inputNames));
    }
    
    public Cars(List<Car> inputs) {
    }
```
주 생성자가 생성자의 마지막에 위치하는 것이 관례임

javajigi commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -1,64 +1,72 @@
 
-    public List<CarSnapshot> toSnapshots() {
-        List<CarSnapshot> snapshots = new ArrayList<>();
+    public List<CarStatus> statuses() {
```
```suggestion
    public RaceResult statuses() {
```
일급 콜렉션이라 이 객체로 반환해도 되지 않을까?

javajigi commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -5,44 +5,33 @@
-        RandomValueMovePolicy movePolicy = () -> false;
 
-        car.move(movePolicy);
+        car.move(condition);
```
```suggestion
       Car car = new Car("자동차");

        car.move(() -> true);
```
위와 같이 구현해도 되지 않을까?

javajigi commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -5,44 +5,33 @@
+        car.move(condition);
 
-        assertThat(car.toSnapshot().distance()).isZero();
+        assertThat(car.status().distance()).isEqualTo(1);
```
```suggestion
        assertThat(car.status()).isEqualTo(new CarStatus("자동차", 1));
```
이와 같이 비교하는 것은?

javajigi reviewed (changes requested):
마지막 단계에 대한 학습 목표를 설정하고, 이 목표에 따라 의식적인 연습을 도전한 모습이 인상적이네요. 💯 
특히 
"하위 객체의 책임이 명확하고 테스트되었다면, 단순 위임은 테스트하지 않는다. 하지만 조건/분기/조합이 있다면 반드시 테스트한다."
이 기준 덕분에 과도한 테스트와 부족한 테스트 사이의 균형을 찾을 수 있었습니다." 
이 경험은 중요하고도 소중한 경험을 했네요.

딱히 피드백할 것은 없었는데요.
그래도 몇 개 눈에 띄는 것이 있어 피드백 했어요.
피드백은 가볍게 반영할 수 있는 것이니 피드백 반영하면서 로또 미션 1단계 병행해 도전해 보시죠.

ghtjr410 committed:
refactor: RandomMovableConditionTest 코드 간결화

RandomMovableConditionTest.java
- 불필요한 변수 제거
- 생성자 호출을 인라인으로 변경해 테스트 코드 단순화

ghtjr410 committed:
feat: Car 동등성 비교 로직 추가 및 테스트 개선

Car.java
- equals/hashCode 오버라이드 추가로 객체 동등성 비교 지원

CarTest.java
- move 테스트를 상태 비교에서 객체 비교로 변경
- 불필요한 MovableCondition 변수 제거 및 인라인 적용
- status 반환 값 검증 테스트 추가

ghtjr410 committed:
refactor: 테스트 코드 단순화 및 동등성 비교 활용

CarTest.java
- status 테스트에서 불필요한 빈 줄 제거

CarsTest.java
- winners 검증을 동등성 비교 한 줄로 단순화
- 불필요한 변수 제거 및 new Winners 비교 방식으로 개선

ghtjr410 committed:
refactor: 생성자 위치 정리 및 클래스 구조 정돈

Car.java
- CarName/CarDistance 기반 생성자 위치 재정렬(가독성 향상)

Cars.java
- inputNames 기반 생성자를 상단으로 이동해 생성자 구조 정리
- 중복되는 생성자 배치 순서 개선

RacingGame.java
- Cars/RacingRound 기반 생성자 위치 재정렬(오버로드 일관성 유지)

RacingRound.java
- 기본 생성자(RacingRound(int))를 상단으로 이동하여 생성자 흐름 개선

ghtjr410 committed:
refactor: 상태 반환 구조를 RaceResult 중심으로 정리

Cars.java
- `statuses()`를 `status()`로 변경
- 기존 List<CarStatus> 반환에서 `RaceResult` 객체 반환으로 통합
- 내부적으로 `new RaceResult(statuses)` 생성하도록 변경

RacingGame.java
- `result()`에서 직접 `new RaceResult(...)` 생성하던 로직을 `cars.status()` 호출로 단순화

ghtjr410 committed:
docs: 리팩터링 문서에 PR 링크 추가

ghtjr410 committed:
docs: 문서 포맷 개선 및 체크리스트 추가

ghtjr410 commented:
# 피드백 반영 사항
- 테스트 단순화 및 동등성 비교 활용
- 생성자 위치 정리
- 하위객체에서 일급콜렉션 반환

리뷰 잘부탁드립니다!

javajigi reviewed (approved):
지금까지 자동차 경주 게임 미션 진행하느라 수고했어요. 👍 
이 미션에서 학습한 내용들을 기반으로 로또 미션 진행해 보시죠.
